### ISSUE TITLE
Differentiate memory bound from left and right subtree

### ISSUE THREAD

**Issue by domainexpert** [tracer-x/klee#298]
Similar to the path condition marking, we should to differentiate memory bound interpolants from the left and right subtrees so that more stricter interpolant does not get applied to the interpolants of the other branch.

**Comment by domainexpert** [comment]
The following example demonstrates some aspect of this problem. Right now, even an exact address interpolant from one branch clobbers the bounded address from another branch resulting in subsumption failures. This is demonstrated by the following program:
```
/*
 * Copyright 2017 National University of Singapore
 *
 * There should be subsumptions among instances of B and C, if
 * slackening actually works. However, the marking due to
 * infeasibility at A is preventing it.
 */
#ifdef LLBMC
#include <llbmc.h>
#else
#include <klee/klee.h>
#endif

int main() {
  int dummy = 0;
  int a[1000];
  int *p = a + 5;
  int ret;
  
  klee_make_symbolic(a, 1000 * sizeof(int), "a");
  
  if (*p) {
    dummy++;
  } else {
    dummy+=3;
    /* A */
    if (*p) {
      dummy++;
    } else {
      dummy += 2;
    }    
  }

  char b;
  klee_make_symbolic(&b, sizeof(char), "b");
  
  if (b) {
    p = a + 4;
    dummy++;
  } else {
    dummy+=2;
  }

  char c;
  klee_make_symbolic(&c, sizeof(char), "c");
  
  if (c) {
    /* B */
    dummy++;
    ret = *p;
  } else {
    /* C */
    ret = *p;
    dummy+=2;
  }

  return ret;
}
```
The expected tree is as follows:
![tree](https://user-images.githubusercontent.com/696558/30854658-e22ceee6-a2e4-11e7-9704-1ed76d236d00.png)
However, due to subsumption loss, the actual tree is the following:
![tree](https://user-images.githubusercontent.com/696558/30854737-240231c8-a2e5-11e7-80b6-2c43634b7dc7.png)

**Comment by domainexpert** [comment]
Joxan suggested an implementation to check if the memory bounds interpolation (slackening) proposed by one branch would open a false path, and if not, we apply the slackening. However, this requires the formulation of checking formula, which involves extracting the symbolic representation of the pointers from memory, which may originally not have symbolic representation, so that the pointer can be represented as a symbolic value in the bound check formula submitted to the solver.

**Comment by domainexpert** [comment]
This issue is resolved via #319.

Note that for the example discussed above, with #319 the tree still has only two subsumptions: This is because the example was wrong: 4, 5, 7, 8 are all descendants of 2, which has unsatisfiability core involving *p, so the subsumption of 7 by 4 and 8 by 5 should not have been possible.
